src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -0,0 +1,253 @@
+}
+
+size_t BasePartialHashIndexImpl::insert_entries(const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>&,
+                                                const ColumnID) {
+  return 0;
+}
+
+size_t BasePartialHashIndexImpl::remove_entries(const std::vector<ChunkID>&) {
+  return 0;
+}
+
+BasePartialHashIndexImpl::Iterator BasePartialHashIndexImpl::cbegin() const {
+  return Iterator(std::make_shared<BaseTableIndexIterator>());
+}
+
+BasePartialHashIndexImpl::Iterator BasePartialHashIndexImpl::cend() const {
+  return Iterator(std::make_shared<BaseTableIndexIterator>());
+}
+
+BasePartialHashIndexImpl::Iterator BasePartialHashIndexImpl::null_cbegin() const {
+  return Iterator(std::make_shared<BaseTableIndexIterator>());
+}
+
+BasePartialHashIndexImpl::Iterator BasePartialHashIndexImpl::null_cend() const {
+  return Iterator(std::make_shared<BaseTableIndexIterator>());
+}
+
+size_t BasePartialHashIndexImpl::memory_usage() const {
+  return 0;
+}
+
+BasePartialHashIndexImpl::IteratorPair BasePartialHashIndexImpl::range_equals(const AllTypeVariant& value) const {
+  return std::make_pair(Iterator(std::make_shared<BaseTableIndexIterator>()),
+                        Iterator(std::make_shared<BaseTableIndexIterator>()));
+}
+
+std::pair<BasePartialHashIndexImpl::IteratorPair, BasePartialHashIndexImpl::IteratorPair>
+BasePartialHashIndexImpl::range_not_equals(const AllTypeVariant& value) const {
+  return std::make_pair(range_equals(value), range_equals(value));
+}
+
+bool BasePartialHashIndexImpl::is_index_for(const ColumnID column_id) const {
+  return false;
+}
+
+std::unordered_set<ChunkID> BasePartialHashIndexImpl::get_indexed_chunk_ids() const {
+  return std::unordered_set<ChunkID>{};
+}
+
+template <typename DataType>
+PartialHashIndexImpl<DataType>::PartialHashIndexImpl(
+    const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id)
+    : BasePartialHashIndexImpl() {
+  insert_entries(chunks_to_index, column_id);
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::insert_entries(
+    const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id) {
+  const size_t size_before = _indexed_chunk_ids.size();
+  for (const auto& chunk : chunks_to_index) {
+    // We do not allow multiple indexing of one chunk.
+    if (_indexed_chunk_ids.contains(chunk.first))
+      continue;
+
+    _indexed_chunk_ids.insert(chunk.first);
+    // Iterate over the segment to index and populate the index.
+    const auto& indexed_segment = chunk.second->get_segment(column_id);
+    segment_iterate<DataType>(*indexed_segment, [&](const auto& position) {
+      const auto row_id = RowID{chunk.first, position.chunk_offset()};
+      // If value is NULL, add to NULL vector, otherwise add into value map.
+      if (position.is_null()) {
+        _null_values.push_back(row_id);
+      } else {
+        if (!_map.contains(position.value())) {
```

Comment:
Doc of `std::map<Key,T,Compare,Allocator>::operator[]`: 
> Returns a reference to the value that is mapped to a key equivalent to key, performing an insertion if such key does not already exist. 

if tsl_map conforms to the STL interface, it's not necessary to insert the empty vector.